autobatch/autobatch.go

```diff
@@ -23,28 +31,79 @@ type op struct {
 }
 
 // NewAutoBatching returns a new datastore that automatically
-// batches writes using the given Batching datastore. The size
-// of the memory pool is given by size.
-func NewAutoBatching(d ds.Batching, size int) *Datastore {
-	return &Datastore{
-		child:            d,
-		buffer:           make(map[ds.Key]op, size),
-		maxBufferEntries: size,
+// batches writes using the given Batching datastore. The maximum number of
+// write before triggering a batch is given by maxWrite. The maximum delay
+// before triggering a batch is given by maxDelay.
+func NewAutoBatching(child ds.Batching, maxWrite int, maxDelay time.Duration) *Datastore {
+	d := &Datastore{
+		child:    child,
+		buffer:   make(map[ds.Key]op, maxWrite),
+		maxWrite: maxWrite,
+		maxDelay: maxDelay,
+		newWrite: make(chan struct{}),
+		exit:     make(chan struct{}),
+	}
+	go d.runBatcher()
+	return d
+}
+
+func (d *Datastore) addOp(key ds.Key, op op) {
+	d.mu.Lock()
+	d.buffer[key] = op
+	d.mu.Unlock()
+	d.newWrite <- struct{}{}
+}
+
+func (d *Datastore) runBatcher() {
+	var timer <-chan time.Time
+
+	write := func() {
+		timer = nil
```

Comment:
This leaks the timer until it fires, e.g. if there is a high rate of ops that hit the maxWrite threshold really fast, we could end up with a lot of concurrent no-op timers running. I think it'd be better to use a proper timer and stop it here.